Don't negotiate TLSv1.3 if our EC cert isn't TLSv1.3 capable

mattcaswell opened the pull request:
TLSv1.3 is more restrictive about the curve used. There must be a matching sig alg defined for that curve. Therefore if we are using some other curve in our certificate then we should not negotiate TLSv1.3.

Additionally we should not apply signature algorithm curve checks in TLSv1.2. RFC8446 says this:
    
        ECDSA signature schemes align with TLS 1.2's ECDSA hash/signature
        pairs.  However, the old semantics did not constrain the signing
        curve.  If TLS 1.2 is negotiated, implementations MUST be prepared
        to accept a signature that uses any curve that they advertised in
        the "supported_groups" extension.

Fixes #7435

kroeckx commented:
Please add tests

mattcaswell commented:
I updated this to add a test and rebased it.

I also dropped the commit about sig alg curve checks in TLSv1.2. This turned out to be unnecessary (and wrong).

kroeckx commented on test/ssl-tests/20-cert-select.conf.in:
```diff
@@ -138,6 +138,28 @@ our @tests = (
+            "ExpectedServerSignType" =>, "EC",
+            # Note: certificate_authorities not sent for TLS < 1.3
+            "ExpectedServerCANames" =>, "empty",
+            "ExpectedResult" => "Success"
```
Maybe ExpectedProtocol = TLSv1.2?

mattcaswell replied:
I thought about that but decided against it on the basis that this test still works if TLSv1.2 is disabled.

kroeckx commented on test/certs/server-ecdsa-brainpoolP256r1-cert.pem:
```diff
@@ -0,0 +1,16 @@
+-----BEGIN CERTIFICATE-----
```
I think we have a script to generate those files. Can you update that?

mattcaswell replied:
Done

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -1515,7 +1516,20 @@ static int is_tls13_capable(const SSL *s)
+        /*
+         * Prior to TLSv1.3 sig algs allowed any curve to be used. TLSv1.3 is
+         * more restrictive so check that our sig algs are consistent with this
+         * EC cert.
```
Can you add a reference?

mattcaswell replied:
Done

kroeckx commented:
I wonder if we should also add some test that fails.

mattcaswell commented:
> I wonder if we should also add some test that fails.

Can you suggest what such a test might look like?

mattcaswell commented:
Ping @kroeckx

mattcaswell commented:
Ping @kroeckx

levitte commented:
> > I wonder if we should also add some test that fails.
> 
> Can you suggest what such a test might look like?

Could that be a cert with an invalid curve for TLSv1.3, perhaps?  I'm not savvy enough to know if that should cause an error or a downgrade to TLSv1.2, but yeah, something like that.  The way I understand this PR, you've only added a test for a cert that successfully enables TLSv1.3, is that correct?

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -1515,7 +1516,20 @@ static int is_tls13_capable(const SSL *s)
+        if (eckey == NULL)
+            continue;
+        curve = EC_GROUP_get_curve_name(EC_KEY_get0_group(eckey));
+        if (tls_check_sigalg_curve(s, curve))
```
I wonder why this check is ECC specific. The requirements seem to be valid for any certificate type.

kroeckx replied:
I guess what I'm proposing is that you make tls_check_sigalg_curve more general.

mattcaswell replied:
I'm not sure I understand. What would a more general version do? The problem we are trying to solve here is that there are 3 sigalgs defined for TLSv1.2 and below that are also used in TLSv1.3 but have a slightly different meaning. These are:

````
          /* ECDSA algorithms */
          ecdsa_secp256r1_sha256(0x0403),
          ecdsa_secp384r1_sha384(0x0503),
          ecdsa_secp521r1_sha512(0x0603),
````
In TLSv1.2 and below this was defined differently based on the combination of the signature algorithm (0x03 for ecdsa) and the hash (0x04 for sha256, 0x05 for sha384 and 0x06 for sha521). So this gives us:

````
ECDSA+SHA256(0x0403)
ECDSA+SHA384(0x0503)
ECDSA+SHA512(0x0603)
````
So the same numbers have different interpretations in the two protocol versions. Personally I think this is probably a mistake in the TLSv1.3 spec. The difference between the two versions is an additional restriction that TLSv1.3 places on these values to include a specific EC curve. Therefore ECDSA certificates that used to work just fine in TLSv1.2 might suddenly stop working in TLSv1.3 if the client doesn't send a sig alg which is consistent with the curve in the certificate. In the specific case of brainpool certs there are no sigalgs defined for those curves at all - so all servers using brainpool curves in their certs will fail to work in TLSv1.3

This problem is inherently EC specific as it only applies to those three sigalgs where the definition changed.

kroeckx replied:
Don't you have the same problem with RSASSA-PSS?

kroeckx replied:
But I guess that's not the same kind of problem, and not something you want to check in this function.

mattcaswell replied:
Exactly.

tomato42 replied:
while RSASSA-PSS is different, isn't DSA in the same boat as brainpool? what's its behaviour with TLS 1.3?

mattcaswell replied:
DSA is already covered by this function (see the beginning of the for loop). If a server is only configured with a DSA cert it will never negotiate TLSv1.3

mattcaswell commented:
> Could that be a cert with an invalid curve for TLSv1.3, perhaps? I'm not savvy enough to know if that should cause an error or a downgrade to TLSv1.2, but yeah, something like that. The way I understand this PR, you've only added a test for a cert that successfully enables TLSv1.3, is that correct?

The tests do add a curve which is invalid for TLSv1.3. Prior to the fix in this PR libssl would have negotiated TLSv1.3 anyway and then subsequently failed. After the fix in this PR libssl realises that its cert is not TLSv1.3 compatibile and successfully negotiates TLSv1.2 instead.

vdukhovni commented on test/certs/server-ecdsa-brainpoolP256r1-key.pem:
```diff
@@ -0,0 +1,5 @@
+MIGIAgEAMBQGByqGSM49AgEGCSskAwMCCAEBBwRtMGsCAQEEIKZSRhbD6lGhKbIm
+5JVgxnN8MHGB0whroUsSf0zmsAz+oUQDQgAETYDLIgpvvoxSBJxB5apcNrTZ0vYp
+VyG18hDEOplqkylnW7kekN9a83WtIwPRoSwhczgFg/MhvLZ/BHQJW2SU3g==
+-----END PRIVATE KEY-----
```
It also makes sense to retrofit the script to generate the key if not already present.  For that, just prefix the makecert.sh command with the right environment variable settings:
```
OPENSSL_KEYALG=ec OPENSSL_KEYBITS=brainpoolP256r1 ./mkcert genee ...
```

mattcaswell replied:
Done

vdukhovni commented on test/ssl-tests/20-cert-select.conf:
```diff
@@ -223,14 +224,43 @@ ExpectedServerSignType = Ed448
+ExpectedResult = Success
+ExpectedServerCANames = empty
+ExpectedServerCertType = brainpoolP256r1
+ExpectedServerSignType = EC
```
Should this also test that the protocol version is TLS 1.2?

mattcaswell replied:
@kroeckx mentioned the same thing. I decided against it on the basis that this test can still work even if TLSv1.2 is disabled.

mattcaswell commented:
@vdukhovni I added a commit to implement your suggestion for the key generation in the script. I also added a commit to add a test which checks for a failed connection if the minimum protocol is TLSv1.3 and brainpool is used as per @kroeckx suggestion.

Can you reconfirm?

@kroeckx are you ok with this now?

kroeckx commented on test/ssl-tests/20-cert-select.conf.in:
```diff
@@ -762,6 +784,23 @@ my @tests_tls_1_3 = (
+        client => {
+            "RequestCAFile" => test_pem("root-cert.pem"),
+            "Groups" => "brainpoolP256r1",
+            "MinProtocol" => "TLSv1.3",
```
I was expecting a test with the server setting this minimum version, but I guess for TLS 1.3 we only send TLS 1.3 is the supported versions and the server will know to fail to. I guess it can be useful to also test that.

kroeckx commented:
Yes, it's fine for me.
